core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java

```diff
@@ -216,65 +215,16 @@ public static XContentType xContentType(InputStream si) throws IOException {
         }
         si.mark(GUESS_HEADER_LENGTH);
         try {
-            final int firstInt = si.read(); // this must be an int since we need to respect the method contract
-            if (firstInt == -1) {
-                return null;
-            }
-
-            final int secondInt = si.read(); // this must be an int since we need to respect the method contract
-            if (secondInt == -1) {
-               return null;
-            }
-            final byte first = (byte) (0xff & firstInt);
-            final byte second = (byte) (0xff & secondInt);
-            if (first == SmileConstants.HEADER_BYTE_1 && second == SmileConstants.HEADER_BYTE_2) {
-                int third = si.read();
-                if (third == SmileConstants.HEADER_BYTE_3) {
-                    return XContentType.SMILE;
-                }
-            }
-            if (first == '{' || second == '{') {
-                return XContentType.JSON;
-            }
-            if (first == '-' && second == '-') {
-                int third = si.read();
-                if (third == '-') {
-                    return XContentType.YAML;
-                }
-            }
-            // CBOR logic similar to CBORFactory#hasCBORFormat
-            if (first == CBORConstants.BYTE_OBJECT_INDEFINITE){
-                return XContentType.CBOR;
-            }
-            if (CBORConstants.hasMajorType(CBORConstants.MAJOR_TYPE_TAG, first)) {
-                // Actually, specific "self-describe tag" is a very good indicator
-                int third = si.read();
-                if (third == -1) {
-                    return null;
-                }
-                if (first == (byte) 0xD9 && second == (byte) 0xD9 && third == (byte) 0xF7) {
-                    return XContentType.CBOR;
-                }
-            }
-            // for small objects, some encoders just encode as major type object, we can safely
-            // say its CBOR since it doesn't contradict SMILE or JSON, and its a last resort
-            if (CBORConstants.hasMajorType(CBORConstants.MAJOR_TYPE_OBJECT, first)) {
-                return XContentType.CBOR;
-            }
-
-            for (int i = 2; i < GUESS_HEADER_LENGTH; i++) {
-                int val = si.read();
-                if (val == -1) {
-                    return null;
-                }
-                if (val == '{') {
-                    return XContentType.JSON;
-                }
-                if (Character.isWhitespace(val) == false) {
+            final byte[] firstBytes = new byte[GUESS_HEADER_LENGTH];
+            int read = 0;
+            while (read < firstBytes.length) {
```

Comment:
I wonder if we should this api to the forbidden api pointing people at the utilities...  (like we did in the write variants).